From 2a951eaa5de3885fa1c5859f8ac08482a1c0bccc Mon Sep 17 00:00:00 2001
From: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Date: Wed, 27 Nov 2024 15:35:53 +0100
Subject: [PATCH 1397/1695] drivers: stm32_stgen: remove support property
 st,tsgen-clk-source

RCC driver configure STGEN flexgen in the .enable ops.
The property st,tsgen-clk-source was used to set the clock source of
stgen_clk. Since we are using the rcc clock tree defined in *-rcc.dtsi
file. We can remove the support of the property from the driver.

Update the probe sequence to stop STGEN before starting the clock
stgen_clk.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 540c572abf103e68527db5e50ecdcb5f2d5df9bf drivers: stm32_stgen: remove support property st,tsgen-clk-source]
Signed-off-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Change-Id: I838fdbbeaa4db2d3d47d33748764eab3a7716077
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/421331
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/arch/arm/dts/stm32mp211.dtsi  |   1 -
 core/arch/arm/dts/stm32mp231.dtsi  |   1 -
 core/arch/arm/dts/stm32mp251.dtsi  |   1 -
 core/drivers/counter/stm32_stgen.c | 100 +++++++----------------------
 4 files changed, 23 insertions(+), 80 deletions(-)

diff --git a/core/arch/arm/dts/stm32mp211.dtsi b/core/arch/arm/dts/stm32mp211.dtsi
index b5f1d22dd..6a459179b 100644
--- a/core/arch/arm/dts/stm32mp211.dtsi
+++ b/core/arch/arm/dts/stm32mp211.dtsi
@@ -2770,7 +2770,6 @@
 			reg = <0x48080000 0x1000>;
 			clocks = <&rcc CK_BUS_STGEN>, <&rcc CK_KER_STGEN>;
 			clock-names = "bus", "stgen_clk";
-			st,tsgen-clk-source = <&rcc HSE_CK>;
 		};
 
 		fmc: memory-controller@48200000 {
diff --git a/core/arch/arm/dts/stm32mp231.dtsi b/core/arch/arm/dts/stm32mp231.dtsi
index fdf1838e1..2df192140 100644
--- a/core/arch/arm/dts/stm32mp231.dtsi
+++ b/core/arch/arm/dts/stm32mp231.dtsi
@@ -3254,7 +3254,6 @@
 			reg = <0x48080000 0x1000>;
 			clocks = <&rcc CK_BUS_STGEN>, <&rcc CK_KER_STGEN>;
 			clock-names = "bus", "stgen_clk";
-			st,tsgen-clk-source = <&rcc HSE_CK>;
 		};
 
 		fmc: memory-controller@48200000 {
diff --git a/core/arch/arm/dts/stm32mp251.dtsi b/core/arch/arm/dts/stm32mp251.dtsi
index 0172f74e1..6d440cb98 100644
--- a/core/arch/arm/dts/stm32mp251.dtsi
+++ b/core/arch/arm/dts/stm32mp251.dtsi
@@ -3403,7 +3403,6 @@
 			reg = <0x48080000 0x1000>;
 			clocks = <&rcc CK_BUS_STGEN>, <&rcc CK_KER_STGEN>;
 			clock-names = "bus", "stgen_clk";
-			st,tsgen-clk-source = <&rcc HSE_CK>;
 		};
 
 		fmc: memory-controller@48200000 {
diff --git a/core/drivers/counter/stm32_stgen.c b/core/drivers/counter/stm32_stgen.c
index 179a88aee..da63cb679 100644
--- a/core/drivers/counter/stm32_stgen.c
+++ b/core/drivers/counter/stm32_stgen.c
@@ -37,7 +37,6 @@ struct stm32_stgen_compat_data {
 
 struct stgen_pdata {
 	struct clk *stgen_clock;
-	struct clk *stgen_clock_source;
 	struct clk *bus_clock;
 	struct stm32_rtc_calendar *calendar;
 	uint32_t cnt_h;
@@ -110,7 +109,7 @@ static void stm32_stgen_pm_suspend(void)
 
 static void stm32_stgen_pm_resume(void)
 {
-	unsigned long clock_src_rate = clk_get_rate(stgen_d.stgen_clock_source);
+	unsigned long clock_src_rate = clk_get_rate(stgen_d.stgen_clock);
 	uint64_t counter_val = ((uint64_t)stgen_d.cnt_h << 32) | stgen_d.cnt_l;
 	struct stm32_rtc_calendar cur_calendar = { };
 	uint64_t nb_pm_count_ticks = 0;
@@ -163,7 +162,6 @@ static TEE_Result parse_dt(const void *fdt, int node)
 	TEE_Result res = TEE_ERROR_GENERIC;
 	const fdt32_t *cuint = NULL;
 	struct io_pa_va base = { };
-	int lenp = 0;
 	size_t reg_size = 0;
 
 	cuint = fdt_getprop(fdt, node, "reg", NULL);
@@ -188,33 +186,16 @@ static TEE_Result parse_dt(const void *fdt, int node)
 	else if (res)
 		panic("No stgen clock available in device tree");
 
-	cuint = fdt_getprop(fdt, node, "st,tsgen-clk-source", &lenp);
-	if (!cuint) {
-		DMSG("No stgen clock source found");
-		return TEE_SUCCESS;
-	}
-
-	lenp /= sizeof(uint32_t);
-	if (lenp != 2)
-		panic("Incorrect stgen clock source");
-
-	stgen_d.stgen_clock_source =
-		stm32mp_rcc_clock_id_to_clk(fdt32_to_cpu(cuint[1]));
-
-	assert(stgen_d.stgen_clock_source);
-
 	return TEE_SUCCESS;
 }
 
 static TEE_Result stgen_probe(const void *fdt, int node,
-			      const void *compat_data)
+			      const void *compat_data __unused)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
 	uint64_t stgen_counter = 0;
-	unsigned long clock_source_rate = 0;
-	unsigned long initial_stgen_rate = 0;
-	struct clk *stgen_source = NULL;
-	const struct stm32_stgen_compat_data *compat_d = compat_data;
+	unsigned long previous_source_rate = 0;
+	unsigned long stgen_rate = 0;
 
 	res = parse_dt(fdt, node);
 	if (res)
@@ -224,35 +205,15 @@ static TEE_Result stgen_probe(const void *fdt, int node,
 	if (!stgen_d.calendar)
 		return TEE_ERROR_OUT_OF_MEMORY;
 
-	/*
-	 * Some SoCs may have intermediate clock signals between stgen and
-	 * desired stgen clock source. Use stgen_source to ease manipulation.
-	 */
-	stgen_source = stgen_d.stgen_clock;
-	if (compat_d->intermediate_clock_parent) {
-		stgen_source = clk_get_parent(stgen_d.stgen_clock);
-		assert(stgen_source);
-	}
-
-	res = clk_enable(stgen_d.stgen_clock);
-	if (res)
-		goto err;
-
 	res = clk_enable(stgen_d.bus_clock);
-	if (res) {
-		clk_disable(stgen_d.stgen_clock);
-		goto err;
-	}
+	if (res)
+		panic();
 
 	/*
-	 * Do nothing if stgen_clock parent clock is already
-	 * targeted clock source or if there is no STGEN source clock
+	 * Read current source rate before configuration of the flexgen to
+	 * update STGEN counter.
 	 */
-	if (!stgen_d.stgen_clock_source ||
-	    clk_get_parent(stgen_source) == stgen_d.stgen_clock_source)
-		goto end;
-
-	initial_stgen_rate = clk_get_rate(stgen_d.stgen_clock);
+	previous_source_rate = clk_get_rate(stgen_d.stgen_clock);
 
 	/*
 	 * Disable STGEN counter. At cold boot, we accept the counter delta
@@ -260,56 +221,41 @@ static TEE_Result stgen_probe(const void *fdt, int node,
 	 */
 	io_clrbits32(stgen_d.base + STGENC_CNTCR, BIT(0));
 
-	clock_source_rate = clk_get_rate(stgen_d.stgen_clock_source);
-
-	res = clk_set_parent(stgen_source, stgen_d.stgen_clock_source);
+	/*
+	 * Flexgen for stgen_clock is skip during RCC probe.
+	 * Enabling stgen_clock configure the flexgen.
+	 */
+	res = clk_enable(stgen_d.stgen_clock);
 	if (res)
-		panic("Couldn't set stgen source");
+		panic();
 
-	res = clk_set_rate(stgen_source, clock_source_rate);
-	if (res)
-		panic("Couldn't set stgen clock rate");
+	stgen_rate = clk_get_rate(stgen_d.stgen_clock);
 
-	io_write32(stgen_d.base + STGENC_CNTFID0, clock_source_rate);
-	if (io_read32(stgen_d.base + STGENC_CNTFID0) != clock_source_rate)
+	io_write32(stgen_d.base + STGENC_CNTFID0, stgen_rate);
+	if (io_read32(stgen_d.base + STGENC_CNTFID0) != stgen_rate)
 		panic("Couldn't modify STGEN clock rate");
 
 	/* Update counter value according to the new STGEN clock frequency */
 	stgen_counter = stm32_stgen_get_counter_value();
-	stgen_counter = stgen_counter * clock_source_rate / initial_stgen_rate;
+	stgen_counter = stgen_counter * stgen_rate / previous_source_rate;
 	set_counter_value(stgen_counter);
 
 	if (IS_ENABLED(CFG_WITH_ARM_TRUSTED_FW))
 		stm32mp_stgen_smc_config();
 
-	register_pm_core_service_cb(stm32_stgen_pm, NULL, "stm32-stgen");
-end:
 	io_setbits32(stgen_d.base + STGENC_CNTCR, BIT(0));
 
-	return TEE_SUCCESS;
-
-err:
-	free(stgen_d.calendar);
+	register_pm_core_service_cb(stm32_stgen_pm, NULL, "stm32-stgen");
 
-	return res;
+	return TEE_SUCCESS;
 }
 
-static const struct stm32_stgen_compat_data stm32_stgen_mp1 = {
-	.intermediate_clock_parent = false,
-};
-
-static const struct stm32_stgen_compat_data stm32_stgen_mp2 = {
-	.intermediate_clock_parent = true,
-};
-
 static const struct dt_device_match stm32_stgen_match_table[] = {
 	{
-		.compatible = "st,stm32mp1-stgen",
-		.compat_data = &stm32_stgen_mp1
+		.compatible = "st,stm32mp1-stgen"
 	},
 	{
-		.compatible = "st,stm32mp25-stgen",
-		.compat_data = &stm32_stgen_mp2
+		.compatible = "st,stm32mp25-stgen"
 	},
 	{ }
 };
-- 
2.47.2

